OptimizelySDK/Odp/OdpEventManager.cs

```diff
@@ -0,0 +1,401 @@
+            _logger.Log(LogLevel.DEBUG, $"Stopped. Queue Count: {_queue.Count}.");
+        }
+
+        /// <summary>
+        /// Associate a full-stack userid with an established VUID
+        /// </summary>
+        /// <param name="userId">Full-stack User ID</param>
+        public void IdentifyUser(string userId)
+        {
+            var identifiers = new Dictionary<string, string>
+            {
+                {
+                    OdpUserKeyType.FS_USER_ID.ToString(), userId
+                },
+            };
+
+            var odpEvent = new OdpEvent(TYPE, "identified", identifiers);
+            SendEvent(odpEvent);
+        }
+
+        /// <summary>
+        /// Send an event to ODP via dispatch queue
+        /// </summary>
+        /// <param name="odpEvent">ODP Event to forward</param>
+        public void SendEvent(OdpEvent odpEvent)
+        {
+            if (State == ExecutionState.Stopped)
+            {
+                _logger.Log(LogLevel.WARN,
+                    "ODP is not enabled.");
+                return;
+            }
+
+            if (!IsOdpConfigurationReady())
+            {
+                return;
+            }
+
+            if (InvalidDataFound(odpEvent.Data))
+            {
+                _logger.Log(LogLevel.ERROR, "ODP data is not valid.");
+                return;
+            }
+
+            odpEvent.Data = AugmentCommonData(odpEvent.Data);
+            Enqueue(odpEvent);
+        }
+
+        /// <summary>
+        /// Add a new ODP event to the queue
+        /// </summary>
+        /// <param name="odpEvent"></param>
+        private void Enqueue(OdpEvent odpEvent)
+        {
+            if (_queue.Count >= _queueSize)
+            {
+                _logger.Log(LogLevel.WARN,
+                    $"ODP event send failed (queueSize = {_queue.Count}).");
+                return;
+            }
+
+            _queue.Enqueue(odpEvent);
+
+            ProcessQueue();
+        }
+
+        /// <summary>
+        /// Process the queue
+        /// </summary>
+        /// <param name="shouldFlush">True if complete flush of queue is needed</param>
+        private void ProcessQueue(bool shouldFlush = false)
+        {
+            if (State != ExecutionState.Running)
+            {
+                return;
+            }
+
+            if (!IsOdpConfigurationReady())
+            {
+                return;
+            }
+
+            _logger.Log(LogLevel.DEBUG,
+                $"Processing Queue {(shouldFlush ? "(flush)" : string.Empty)}");
+
+            if (shouldFlush)
+            {
+                ClearCurrentTimeout();
+
+                State = ExecutionState.Processing;
+
+                while (QueueContainsItems())
+                {
+                    MakeAndSend1Batch();
+                }
+            }
+            else if (QueueHasBatches())
+            {
+                ClearCurrentTimeout();
+
+                State = ExecutionState.Processing;
```

Comment:
It looks like this State switch cannot protect from multi threads sending at the same time. We need to review this process again.

Reply:
Yes I agree. There's definitely work to be done to figure out threading correctly.  I need to carefully review @zashraf1985 's Java, and see how he & that language handle fully cancelling a timer in progress.   This is hard stuff. Happy for any advice from you and @msohailhussain

Reply:
@mikechu-optimizely you can also look at how it's implemented in the current event dispatcher, which shares the same logic for batching.

Reply:
Will do.  I believe @msohailhussain is reviewing today, and we'll link up Monday over his eval.

Reply:
@jaeopt is the updated batching logic better?